File: src/org/cerion/symcalc/expression/function/arithmetic/Plus.java
package org.cerion.symcalc.expression.function.arithmetic;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.NumberExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

import java.util.Iterator;
import java.util.List;

public class Plus extends FunctionExpr {

	public Plus(Expr... e) {
		super(FunctionType.PLUS,e);
	}
	
	@Override
	protected Expr evaluate() {
		
		NumberExpr sum = IntegerNum.ZERO;
		List<Expr> list = getArgs();

		Iterator<Expr> it = list.iterator();
		while(it.hasNext())
		{
			Expr e = it.next();
			if(e.isNumber())
			{
				sum = sum.add((NumberExpr)e);
				it.remove();
			}
		}

		//At least 1 non-number value
		if(list.size() > 0)
		{
			//Only add number if non-zero
			if(!sum.isZero())
				list.add(0, sum);

			//If only 1 entry just return it
			if(list.size() == 1)
				return list.get(0);
			
			return new Plus(list.toArray(new Expr[list.size()]));
		}
		
		return sum;
	}
	
	@Override
	public String toString() {
		if(size() > 0)
		{
			String s = get(0).toString();
			for(int i = 1; i < size(); i++)
				s += " + " + get(i).toString();
			
			return s;
		}
		
		return super.toString();
	}

	@Override
	protected int getProperties() {
		return Properties.ASSOCIATIVE.value;
	}
}

File: src/main/kotlin/function/arithmetic/Plus.kt
package org.cerion.symcalc.function.arithmetic

import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.constant.ComplexInfinity
import org.cerion.symcalc.function.FunctionExpr
import org.cerion.symcalc.expression.number.Integer
import org.cerion.symcalc.expression.number.NumberExpr
import java.util.*

class Plus(vararg e: Any) : FunctionExpr(*e) {

    override val properties: Int
        get() = Properties.Flat.value or Properties.NumericFunction.value or Properties.Orderless.value or Properties.LISTABLE.value

    override fun evaluate(): Expr {
        val list = ArrayList<Expr>()
        for (i in 0 until size)
            list.add(get(i))

        // Combine number values
        val numberItems = list.filterIsInstance<NumberExpr>()
        list.removeIf { it is NumberExpr }
        val sum = numberItems.fold(Integer.ZERO as NumberExpr) { acc, n -> acc + n }
        if (!sum.isZero)
            list.add(sum)

        //At least 1 non-number value
        if (list.size > 0) {
            val groups = list.groupBy { it.toString() }
            for (group in groups) {
                if (group.value.size > 1) {
                    list.removeIf { it.toString() == group.key }
                    list.add(Times(Integer(group.value.size), group.value.first()))
                }
            }

            // If element is also contained in Times(), remove it and increment by 1  (2*Pi) + Pi = 3*Pi
            val timesArg = list.filter { it is Times && it.size == 2 && it[0] is NumberExpr && it[1] !is NumberExpr }
            for (i in 0 until timesArg.size) {
                val times = timesArg[i] as Times
                while (list.contains(times[1])) {
                    val t = (times[0] + Integer.ONE) * times[1]
                    list.remove(times)
                    list.remove(times[1])
                    list.add(t)
                }
            }

            //If only 1 entry just return it
            if (list.size == 1)
                return list[0]

            if (list.contains(ComplexInfinity()))
                return ComplexInfinity()

            return Plus(*list.toTypedArray())
        }

        return sum
    }

    override fun toString(): String {
        if (size > 0) {
            val strings = args.map {
                if (it is Times)
                    "($it)"
                else
                    it.toString()
            }

            return strings.joinToString(" + ")
        }

        return super.toString()
    }

    override fun validate() {
    }
}
